backend/tracim_backend/lib/utils/authentication.py

```diff
@@ -20,16 +28,195 @@
+        self._metadata_response = Response(content_type='application/xml')
+        self._metadata_response.text = create_metadata_string(config=self.saml_config,
+                                                              configfile=None).decode()
+        self.saml_client = Saml2Client(config=self.saml_config)
+
+        for _, data in self.saml_client.config.vorg.items():
+            if data.common_identifier not in app_config.IDP_LIST:
+                app_config.IDP_LIST.append(data.common_identifier)
+
+        configurator.add_forbidden_view(self._idp_chooser)
+        configurator.add_route("acs", "/saml/acs", request_method="POST")
+        configurator.add_route("sso", "/saml/sso", request_method="GET")
+        configurator.add_route("metadata", "/saml/metadata", request_method="GET")
+        configurator.add_view(self._acs, route_name="acs")
+        configurator.add_view(self._sso, route_name="sso")
+        configurator.add_view(self._metadata, route_name="metadata")
+
+    def _load_settings(self, config: Configurator):
+        # TODO - M.L - 2023/09/05 - Change this to proper config loading
+        """Parse and validate SAML configuration.
+
+        Args:
+            config (pyramid.config.Configurator): The configuration of the
+                existing Pyramid application.
+        """
+
+        if 'PYRAMID_SAML_PATH' in os.environ:
+            if 'pyramid_saml' not in config.get_settings():
+                config.get_settings().update({
+                    'pyramid_saml': {}
+                })
+            config.get_settings().get('pyramid_saml').update({
+                'saml_path': os.environ.get('PYRAMID_SAML_PATH')
+            })
+
+        if 'pyramid_saml' not in config.get_settings():
+            msg = 'Missing \'pyramid_saml\' configuration in settings'
+            raise AssertionError(msg)
+        else:
+            settings = config.get_settings().get('pyramid_saml')
+
+        if 'saml_path' not in settings:
+            msg = 'Missing \'saml_path\' in settings'
+            raise AssertionError(msg)
+
+    def authenticated_userid(
+        self, request: TracimRequest
+    ) -> typing.Optional[typing.Union[str, int]]:
+        # FIXME - SGD - 2023-07-26 - Return Tracim user id when SAML data is present in the
+        #  session. By matching it with SAML ID attribute (in external_id,
+        #  see https://github.com/tracim/tracim/issues/6209)
+        saml_user_id = request.session.get("saml_user_id")
+        if saml_user_id is None:
+            return None
+
+        user_api = UserApi(None, request.dbsession, self.app_config)
+
+        user = user_api.saml_authenticate(
+            user=None,
+            user_id=saml_user_id,
+            name=request.session.get("saml_name"),
+            mail=request.session.get("saml_mail"),
+        )
+        self.remember(request, user.user_id)
+        request.set_user(user)
+        return user.user_id
```

Comment:
This function (which seems to be called each time to check the session), checks the session data and calls the `saml_authenticate` function, which will try to find the user with the `user_id` and if it does not find it, creates it.